api/app/models/comment.rb

```diff
@@ -28,9 +29,63 @@ class Comment < ApplicationRecord
                       counter_cache: :children_count
   has_many :children, class_name: "Comment", foreign_key: :parent_id, dependent: :destroy,
                       inverse_of: :parent, counter_cache: :children_count
-  has_many :flags, as: :flaggable, dependent: :destroy, inverse_of: :flaggable
+  has_one :comment_created_event, -> { where event_type: Event::COMMENT_CREATED },
+          class_name: Event,
+          as: :subject,
+          dependent: :destroy,
+          inverse_of: :subject
+  delegate :project, to: :subject
 
   # Validations
   validates :body, :subject, presence: true
 
+  # Callbacks
+  after_create :notify_parent_creator!, :notify_project_editors!
+  after_commit :trigger_event_creation, on: [:create]
+
+  # rubocop:disable Metrics/LineLength, Metrics/AbcSize
+  def subject_url
+    client_url = Rails.configuration.manifold.url
+
+    if subject.is_a? Resource
+      "#{client_url}/projects/#{subject.project.slug}/resource/#{subject.slug}"
+    elsif subject.is_a? Annotation
+      "#{client_url}/read/#{subject.text.slug}/section/#{subject.text_section_id}#annotation-#{subject.id}"
+    end
+  end
+  # rubocop:enable Metrics/LineLength, Metrics/AbcSize
+
+  private
+
+  def trigger_event_creation
+    Event.trigger(Event::COMMENT_CREATED, self)
+  end
+
+  def notify_parent_creator!
+    return unless parent.present?
+    return if parent.creator == creator
+    return unless parent.creator.wants_notifications_for?("replies_to_me")
+
+    NotificationMailer.reply_notification(parent.creator, self).deliver
+  end
+
+  # rubocop:disable Metrics/AbcSize
+  def notify_project_editors!
```

Comment:
The same problem flagged above in `notify_parent_creator!` applies here. This method is doing work that it shouldn't be. Figuring out who gets notified is not a concern of the comment model. Put this logic in a job, or a service class.

Reply:
✔